From 7c4ba6b420a6c353f0e1f9675205219db2ba2dd0 Mon Sep 17 00:00:00 2001
From: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
Date: Fri, 21 Jul 2023 13:04:17 -0700
Subject: [PATCH] wpa_supplicant: deauth on receiving low ack from AP

Currently only AP and TDLS STA are handled for low ack event from
kernel. Add support for STA mode.

Kernel currently sends NL80211_ATTR_CQM_PKT_LOSS_EVENT when it sees
STA is not reachable due to beacon loss or underlying hardware reports
STA_KICKOUT. Use this event in supplicant to drop the connection for
STA mode.

Signed-off-by: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
---
 wpa_supplicant/events.c | 4 ++++
 1 file changed, 4 insertions(+)

--- a/wpa_supplicant/events.c
+++ b/wpa_supplicant/events.c
@@ -6148,6 +6148,13 @@ void supplicant_event(void *ctx, enum wp
 			wpa_tdls_disable_unreachable_link(wpa_s->wpa,
 							  data->low_ack.addr);
 #endif /* CONFIG_TDLS */
+#ifdef CONFIG_MESH
+		if (wpa_s->ifmsh) {
+			if (data->low_ack.num_packets == 0xFFFF)
+				wpas_mesh_peer_remove(wpa_s, data->low_ack.addr);
+		} else
+#endif
+			wpa_supplicant_deauthenticate(wpa_s, WLAN_REASON_DISASSOC_LOW_ACK);
 		break;
 	case EVENT_IBSS_PEER_LOST:
 #ifdef CONFIG_IBSS_RSN
